### PR TITLE
`kinematics.spin` and `kinematics.eddy_kinetic_energy`

### PR DESCRIPTION
**Pull request by milancurcic**
* [x] `eddy_kinetic_energy`:
  - [x] implementation
  - [x] docs
  - [x] tests
* [x] `spin`:
  - [x] implementation
  - [x] docs
  - [x] tests

### CODE REVIEW

**Comment by selipot** [comment]
The formula reference should be in [Griffa et al. 2008](https://agupubs.onlinelibrary.wiley.com/doi/full/10.1029/2007GL032100)

**Comment by milancurcic** [comment]
@selipot take a look at the current implementation.

File: clouddrift/kinematics.py [hunk 1]
```diff
@@ -19,6 +20,62 @@
...
+    """Compute eddy kinetic energy from zonal and meridional velocities.
+
+    Eddy kinetic energy is defined as one half the variance of the velocity field.
+
+    To obtain the result, variances along ``time_axis`` are computed for each
+    velocity component. The rank of the result is thus reduced by one relative
+    to the input arrays.
+
+    The eddy perturbations are relative to the time-mean velocity, so if you
+    need to apply a shorter time scale, you need to first low-pass filter
```
Comment (selipot) [inline clouddrift/kinematics.py:37]: I don't really understand this paragraph. Could you rephrase?

File: clouddrift/kinematics.py [hunk 2]
```diff
@@ -667,3 +724,147 @@ def velocity_from_position(
     dy /= dt
 
     return np.swapaxes(dx, time_axis, -1), np.swapaxes(dy, time_axis, -1)
+
+
+def spin(
```
Comment (selipot) [inline clouddrift/kinematics.py:729]: Are we sorting alphabetically functions by convention? Just wondering.

File: clouddrift/kinematics.py [hunk 3]
```diff
@@ -667,3 +724,147 @@ def velocity_from_position(
...
+
+    Forward and backward schemes are effectively the same except that the
+    position at which the velocity is evaluated is shifted one element down in
+    the backward scheme relative to the forward scheme. In the case of a
+    forward or backward difference scheme, the last or first element of the
+    velocity, respectively, is extrapolated from its neighboring point. In the
+    case of a centered difference scheme, the start and end boundary points are
+    evaluated using the forward and backward difference scheme, respectively.
+
+    Reference for the definition of spin is:
```
Comment (selipot) [inline clouddrift/kinematics.py:763]: Put instead in a References section as in:

References
----------
.. [1] Veneziani, M., Griffa, A., Reynolds, A.M., Garraffo, Z.D. and Chassignet,
    E.P., 2005. Parameterizations of Lagrangian spin statistics and particle
    dispersion in the presence of coherent vortices. Journal of Marine Research,
    63(6), pp.1057-1083.
.. [2] if there is another one

File: clouddrift/kinematics.py [hunk 4]
```diff
@@ -667,3 +724,147 @@ def velocity_from_position(
...
+
+    Veneziani, M., Griffa, A., Reynolds, A.M., Garraffo, Z.D. and Chassignet,
+    E.P., 2005. Parameterizations of Lagrangian spin statistics and particle
+    dispersion in the presence of coherent vortices. Journal of Marine Research,
+    63(6), pp.1057-1083.
+
+    Parameters
+    ----------
+    u : np.ndarray
+        Zonal velocity
```
Comment (selipot) [inline clouddrift/kinematics.py:773]: I always end with a period as in
    Zonal velocities.

Should we have a convention?

File: clouddrift/kinematics.py [hunk 5]
```diff
@@ -667,3 +724,147 @@ def velocity_from_position(
...
+    # np.swapaxes returns a view of the array, not a copy, if the input is a
+    # NumPy array. Otherwise, it returns a copy.
+    u = np.swapaxes(u, time_axis, -1)
+    v = np.swapaxes(v, time_axis, -1)
+    time = np.swapaxes(time, time_axis, -1)
+
+    if not time.shape == u.shape:
+        # time is 1-d array; broadcast to u.shape.
+        time = np.broadcast_to(time, u.shape)
+
```
Comment (selipot) [inline clouddrift/kinematics.py:825]: Make this an option with a keyword argument set to False by default such as `anomaly=False` i.e. by default the input velocities are not anomalies. I am thinking of an implementation for which a mean spatial field is removed first from the velocities. In that case the user passes `anomaly = True`.

**Review by selipot — CHANGES_REQUESTED** [review]

File: clouddrift/kinematics.py (hunk 4 above)
Comment (milancurcic) [reply clouddrift/kinematics.py:773]: The pedantically correct thing to do (as I learned from my book editor) is to use a period when we have a full sentence (e.g. "This is a zonal velocity.") and to omit it when it's not a sentence ("Zonal velocities").

We don't need to do that because almost nobody will care. We can use periods always if you prefer it.

File: clouddrift/kinematics.py (hunk 5 above)
Comment (milancurcic) [reply clouddrift/kinematics.py:862]: Does `anomaly=False` mean that we compute KE and spin based on full velocities (as input)? Likewise, if `anomaly=True`, we compute EKE by demeaning the input velocities first (current implementation)?

Or, do you mean that `anomaly=False` is a way for the user to communicate whether the input velocity is a full velocity or an anomaly and we demean or not based on that information?

Comment (selipot) [reply clouddrift/kinematics.py:856]: `anomaly=False` means the input velocities are not anomalies and hence we estimate the mean first as the trajectory mean in order to calculate the velocity anomalies and thus the spin and the eke.

Comment (milancurcic) [reply clouddrift/kinematics.py:856]: OK. And what do we do when `anomaly=True`?

Comment (selipot) [reply clouddrift/kinematics.py:856]: `anomalies=True` means the input velocity components are already the u' and v' of your formula. 

Which makes me think that the `eddy_kinetic_energy` function should probably be replaced by a `kinetic_energy` function which calculates `np.mean(u**2) + np.mean(v**2)`. If the input is an anomaly it is an EKE, if the input is not an anomaly it is a total kinetic energy.

Comment (selipot) [reply clouddrift/kinematics.py:856]: That's an option which I support. In this case the code should not remove the Lagrangian mean from the input.

**Comment by selipot** [comment]
Add DOIs to references?

File: clouddrift/kinematics.py [hunk 6]
```diff
@@ -704,3 +745,186 @@ def velocity_from_position(
...
+
+def spin(
+    u: np.ndarray,
+    v: np.ndarray,
+    time: np.ndarray,
+    difference_scheme: Optional[str] = "forward",
+    time_axis: Optional[int] = -1,
+) -> Union[float, np.ndarray]:
+    """Compute spin continuously from velocities and times.
```
Comment (milancurcic) [inline clouddrift/kinematics.py:757]: @selipot what do you mean by "continuously" here? I find it more confusing than without this word.

Comment (selipot) [reply clouddrift/kinematics.py:757]: Continuous because it does not reduce the dimension of the input. Spin studies typically calculate an average of the spin over a segment length. So it is usually a single number for a trajectory or a segment. Could be removed.

Comment (milancurcic) [reply clouddrift/kinematics.py:757]: In the current implementation we reduce over `time_axis`. Should we not?

Comment (milancurcic) [reply clouddrift/kinematics.py:757]: On second read of Veneziani et al. (2005), eq. 2, there should be no time averaging. I confused the $< >$ brackets for a time average (it's a trajectory ensemble average). This will further simplify the code.

**Comment by selipot** [comment]
Another implementation @RickLumpkin reminded me of https://agupubs.onlinelibrary.wiley.com/doi/10.1002/2015JC011435

**Comment by milancurcic** [comment]
This is now ready for another review. It bumps the version to 0.26.0.

**Review by selipot — APPROVED** [review]
